api/envoy/service/mutation/v3alpha/mutation.proto

```diff
@@ -0,0 +1,302 @@
+// single Body message downstream containing the entire message body.
+// It is still possible for the server to modify the message headers
+// with the next upstream message.
+message SendBufferedRequestBodyCommand {
+}
+
+// Request that Envoy buffer and send the response body in a single chunk.
+message SendBufferedResponseBodyCommand {
+}
+
+// Request that Envoy will send the HTTP request trailers when they are
+// available. Once trailers have been requested, no more bodies
+// will be delivered, so this must be sent after all desired body chunks are
+// desired.
+message SendRequestTrailersCommand {
+}
+
+// Request that Envoy send the response trailers when available.
+message SendResponseTrailersCommand {
+}
+
+// The mutation server may respond with zero or more of these responses
+// before closing the stream.
+// [#next-free-field: 16]
+message MutationResponse {
+  oneof command {
+    option (validate.required) = true;
+
+    StopCommand stop = 1;
+
+    ContinueRequestCommand continue_request = 2;
+
+    ContinueResponseCommand continue_response = 3;
+
+    SendRequestBodyChunkCommand send_request_body_chunk = 4;
+
+    SendResponseBodyChunkCommand send_response_body_chunk = 5;
+
+    SendBufferedRequestBodyCommand send_buffered_request_body = 6;
+
+    SendBufferedResponseBodyCommand send_buffered_response_body = 7;
+
+    SendRequestTrailersCommand send_request_trailers = 8;
+
+    SendResponseTrailersCommand send_response_trailers = 9;
+  }
+
+  // Replace the entire HTTP response (whether on the request or on the
+  // response path) with the contents of the CompleteResponse message.
+  // If a StopCommand command is set, then send this response back immediately
+  // to the client with no additional filter processing.
+  CompleteResponse complete_response = 10;
+
+  // Add or replace HTTP headers. This will be ignored after SendXXXBodyChunkCommand has
+  // been sent. If SendXXXTrailers was previously sent, then
+  // modify the trailers instead.
+  repeated config.core.v3.HeaderValueOption set_headers = 11;
+
+  // Remove these HTTP headers. This will be ignored after SendXXXBodyChunk
+  // has been sent. If SendXXXTrailers was previously sent, then
+  // modify the trailers instead.
+  repeated string remove_headers = 12;
+
+  // Replace some or all of the last message body received with a new
+  // set of bytes. This will be ignored unless the previous message sent
+  // by the client is a "Body" request.
+  BodyMutation body_mutation = 13;
+
+  // Add new metadata to the Envoy context
+  google.protobuf.Struct new_metadata = 14;
+
+  // Change how and if the filter waits for responses from the server
+  ProcessingModeChange async_mode_change = 15;
+}
+
+// This message is sent to the mutation server when the HTTP request is
+// first received.
+// [#next-free-field: 9]
+message Request {
+  // A unique ID that will match the one sent in the "Response" message
+  string id = 1;
```

Comment:
Is this needed given that all of this happens on a single bi-di stream?

Reply:
Nope - I will remove it.